src/analysis/schwartz.lean

```diff
@@ -0,0 +1,400 @@
+end
+
+end aux
+
+/-! ### Algebraic properties -/
+
+section smul
+
+variables [normed_space ℂ F]
+variables [semiring R] [module R ℂ] [module R F] [smul_comm_class ℝ R F]
+variables [has_continuous_const_smul R F] [is_scalar_tower R ℂ F]
+variables [semiring R'] [module R' ℂ] [module R' F] [smul_comm_class ℝ R' F]
+variables [has_continuous_const_smul R' F] [is_scalar_tower R' ℂ F]
+
+-- Note that we define the scalar multiplication only in the case that `F` is a vector space
+-- over `ℂ`. The reason for this is that the type-system cannot infer instances if we were to
+-- replace `ℂ` by `[is_R_or_C 𝕜]`. This is mathemically no problem, because the usual Schwartz
+-- space is `𝓢(E, ℂ)` and the space `𝓢(E, ℝ)` is never used in mathematics.
+instance : has_smul R 𝓢(E, F) :=
+⟨λ c f, { to_fun := c • f,
+  smooth' := f.smooth.const_smul c,
+  decay' := λ k n, begin
+    rcases f.decay k n with ⟨C, hC, hf⟩,
+    refine ⟨C * (∥c • (1 : ℂ)∥+1), by positivity, _⟩,
+    intros x,
+    specialize hf x,
+    have hc : 0 ≤ ∥c • (1 : ℂ)∥ := by positivity,
+    refine le_trans _ ((mul_le_mul_of_nonneg_right hf hc).trans _),
+    { refine eq.le _,
+      nth_rewrite 1 mul_comm,
+      rw ←mul_assoc,
+      refine decay_smul_aux k n f c x },
+    rw [mul_le_mul_left hC, le_add_iff_nonneg_right],
+    exact zero_le_one,
+  end}⟩
+
+@[simp] lemma smul_apply {f : 𝓢(E, F)} {c : R} {x : E} : (c • f) x = c • (f x) := rfl
+
+instance [has_smul R R'] [is_scalar_tower R R' F] : is_scalar_tower R R' 𝓢(E, F) :=
+⟨λ a b f, ext $ λ x, smul_assoc a b (f x)⟩
+
+instance [smul_comm_class R R' F] : smul_comm_class R R' 𝓢(E, F) :=
+⟨λ a b f, ext $ λ x, smul_comm a b (f x)⟩
+
+end smul
+
+section zero
+
+instance : has_zero 𝓢(E, F) :=
+⟨{ to_fun := λ _, 0,
+  smooth' := cont_diff_const,
+  decay' := λ _ _, ⟨1, zero_lt_one, λ _, by simp⟩ }⟩
+
+instance : inhabited 𝓢(E, F) := ⟨0⟩
+
+lemma coe_zero : ↑(0 : 𝓢(E, F)) = (0 : E → F) := rfl
+
+@[simp] lemma coe_fn_zero : coe_fn (0 : 𝓢(E, F)) = (0 : E → F) := rfl
+
+@[simp] lemma zero_apply {x : E} : (0 : 𝓢(E, F)) x = 0 := rfl
+
+end zero
+
+section neg
+
+instance : has_neg 𝓢(E, F) :=
+⟨λ f, ⟨-f, f.smooth.neg,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨C, hC, hf⟩,
+    use [C, hC],
+    intro x,
+    exact (decay_neg_aux k n f x).le.trans (hf x),
+  end⟩ ⟩
+
+end neg
+
+section add
+
+instance : has_add 𝓢(E, F) :=
+⟨λ f g, ⟨f + g, f.smooth.add g.smooth,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨Cf, hCf, hf⟩,
+    rcases g.decay k n with ⟨Cg, hCg, hg⟩,
+    refine ⟨Cf + Cg, by positivity, λ x, _⟩,
+    exact (decay_add_le_aux k n f g x).trans (add_le_add (hf x) (hg x)),
+  end⟩ ⟩
+
+@[simp] lemma add_apply {f g : 𝓢(E, F)} {x : E} : (f + g) x = f x + g x := rfl
+
+end add
+
+section sub
+
+instance : has_sub 𝓢(E, F) :=
+⟨λ f g, ⟨f - g, f.smooth.sub g.smooth,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨Cf, hCf, hf⟩,
+    rcases g.decay k n with ⟨Cg, hCg, hg⟩,
```

Comment:
all these proofs would be nicer to write if `seminorm_aux` were already available. Do you think you could move this definition and its basic properties earlier in the file, to avoid all the rcases?

Reply:
I don't really understand what you are suggesting: to use `seminorm_aux` you need to know that `f + g` is a Schwartz function already and this is what we want to prove. It surely is a bit redundant, but I see now easy way around this.

Reply:
You can move earlier the definition and the first properties of `seminorm_aux` (but not all of them, for sure -- you could intersperse the current lemmas and the `seminorm_aux` implications, for instance).